File: processors/message_processor.py
import logging
from services.arcade_services import get_thread_message_list
from services.lm_services import call_lm
from utils.utils import markdown_to_json
from entities.message import Message
from services.arcade_services import change_email_labels
from entities.labels import HoloformLabel, GmailSystemLabel

# Initialize logger for the current module
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)  # Set the logging level to DEBUG

# Create a formatter with the desired log message format
formatter = logging.Formatter(
    '\n%(asctime)s - %(name)s - %(levelname)s - %(message)s\n'
)

# Create a stream handler to output logs to the console
stream_handler = logging.StreamHandler()
stream_handler.setFormatter(formatter)  # Assign the formatter to the handler

# Add the configured handler to the logger
logger.addHandler(stream_handler)

def create_message_objects(gmail_messages: list, user_id: str) -> list[Message]:
    """
    Converts a list of Gmail message dictionaries into Message objects.

    Args:
        gmail_messages: List of Gmail message dictionaries.
        user_id: ID of the user owning the thread
    Returns:
        List of Message objects.
    """
    if not gmail_messages:
        logger.info("No Gmail messages found.")
        return []

    messages = []
    for message in gmail_messages:
        if "id" not in message:
            continue

        logger.debug(f"Creating Message object for Message: {message['id']}")

        message_obj = Message(
            user_id=user_id,
            gmail_message_id=message["id"],
            header_message_id=message["header_message_id"],
            thread_id=message["thread_id"],
            labels=message["label_ids"],
            snippet=message["snippet"],
            history_id=message["history_id"],
            to=message["to"],
            cc=message["cc"],
            sender=message["from"],
            reply_to_address=message["reply_to"],
            date=message["date"],
            subject=message["subject"],
            body=message["body"],
            in_reply_to=message["in_reply_to"],
            references=message["references"],

        )

        logger.debug(f"message_obj created: {message_obj.gmail_message_id}")
        messages.append(message_obj)

    return messages

def summarize_thread(message: Message) -> None:
    prompt = (f"{message} \n"
                f"Summarize the following thread of emails. "
                f"Put it in 2nd person to {message.user_id} and highlight the most important parts to them. "
                f"And if anyone is waiting on them for a response, highlight that prominently. "
                f"Be concise, clear, direct, specific, and correct."
                )
        
    # Summarize the thread
    message.summary = call_lm(prompt=prompt)

def classify_message(message: Message) -> None:
    prompt = f"""
        ## Message: {message}
        ----
        ## Prompt:
        Classify the email into one of the following categories with definitions:
        {HoloformLabel.get_classification_definitions()}

        Then assign a priority level:
        {HoloformLabel.get_priority_definitions()}

        Consider:
        - **Content:** Analyze the main topics and purpose.
        - **Tone:** Determine the formality and sentiment.
        - **Sender:** Identify if the sender is known or reputable.
        - **Time-Sensitivity:** Assess if immediate action is needed.
        - **Calls to Action:** Look for requests or required responses.


        RESPONSE FORMAT:
        Return JSON only: {{"classification": "...", "priority": "...", "explanation": "..."}}
        ----"""
    response = call_lm(prompt=prompt)
    classification = markdown_to_json(response)

    # Add the Holoform prefix to the classification and priority
    classification["classification"] = HoloformLabel.get_full_label(classification["classification"])
    classification["priority"] = HoloformLabel.get_full_label(classification["priority"])
    # Assign the classification to the message
    message.classification = classification

    logger.debug(f"\nMessage Classification: {classification['classification']}\n")
    logger.debug(f"Priority Level: {classification['priority']}\n")
    logger.debug(f"Explanation: {classification['explanation']}\n")

def update_labels(message: Message) -> None:
    logger.debug(f"Updating labels for message: {message.gmail_message_id}")
    labels_to_add = [HoloformLabel.HOLOFORM.full_label, message.classification["classification"], message.classification["priority"]]
    labels_to_remove = []
    if not (message.classification["classification"].lower() == HoloformLabel.ACTION_REQUIRED.full_label or
            message.classification["priority"] == HoloformLabel.HIGH_PRIORITY.full_label):
        labels_to_remove = [GmailSystemLabel.INBOX.value]
    logger.debug(f"\nLabels to add: {labels_to_add}\n")
    logger.debug(f"\nLabels to remove: {labels_to_remove}\n")
    change_email_labels(
        email_id=message.gmail_message_id,
        labels_to_add=labels_to_add,
        labels_to_remove=labels_to_remove,
        user_id=message.user_id,
    )
    logger.debug(f"Finished updating labels for message: {message.gmail_message_id}\n")

def draft_response(message: Message) -> None:
    draft_prompt = (
        f"## Message: {message}\n"
        f"{'-'*8}\n"
        "## Prompt: \n"
        "Draft a response to the message. "
        f"Write in the first person as {message.user_id}. "
        "Be concise, clear, direct, specific, authentic, empathetic, personal, and correct. "
        "Use technical language where appropriate but remain plainspoken. Avoid clichés. "
        "The response should be a reply-all, sent to the sender and all other recipients. Retain CC. "
        "Reponse should be in JSON format: {{'thread_id': '...', 'to': '...', 'cc': '...', 'subject': '...', 'body': '...'}}"
    )
    draft_response_email = (f"response draft: {call_lm(prompt=draft_prompt, tools=['Google.WriteDraftResponseEmail'], tool_choice="Generate")} \n")
    logger.debug(f"draft_response_email: {draft_response_email}\n")

def process_message(message: Message) -> Message:
    """
    Process a single message
    """
    message.thread_messages = get_thread_message_list(message.thread_id, message.user_id)
    classify_message(message)
    logger.debug(f"\n\n Subject: {message.subject} \n Classification: {message.classification} \n {'-'*20}\n")
    update_labels(message)
    logger.debug(f"\n\nFinished processing message: {message.gmail_message_id}\n\n")
    message.status = "processed"
    return message

def process_messages(messages: list[Message]) -> list[Message]:
    return [process_message(message) for message in messages]

File: src/processors/message-processor.ts
import {
  getClassificationDefinitions,
  getFullLabel,
  getPriorityDefinitions,
  GmailSystemLabel,
  HoloformLabel,
} from '../entities/labels.ts';
import { type Classification, Message } from '../entities/message.ts';
import {
  changeEmailLabels,
  type FetchedEmail,
  getThreadMessageList,
} from '../services/arcade-service.ts';
import { callLanguageModel } from '../services/language-model.ts';
import { logger } from '../utilities/logger.ts';
import { markdownToJson } from '../utilities/markdown.ts';

/** Converts a list of fetched emails into {@link Message} objects. */
export function createMessageObjects(emails: FetchedEmail[], userId: string): Message[] {
  if (emails.length === 0) {
    logger.info('No Gmail messages found.');
    return [];
  }

  const messages: Message[] = [];
  for (const email of emails) {
    logger.debug(`Creating Message object for Message: ${email.id}`);
    const message = new Message({
      userId,
      gmailMessageId: email.id,
      headerMessageId: email.headerMessageId ?? '',
      threadId: email.threadId ?? '',
      labels: email.labelIds,
      snippet: email.snippet,
      historyId: email.historyId,
      to: email.to,
      cc: email.cc,
      sender: email.from,
      replyToAddress: email.replyTo,
      date: email.date,
      subject: email.subject,
      body: email.body,
      inReplyTo: email.inReplyTo,
      references: email.references,
    });
    logger.debug(`message_obj created: ${message.gmailMessageId}`);
    messages.push(message);
  }
  return messages;
}

/** Summarizes the message's thread into the message's `summary` field. */
export async function summarizeThread(message: Message): Promise<void> {
  const prompt =
    `${JSON.stringify(message)} \n` +
    'Summarize the following thread of emails. ' +
    `Put it in 2nd person to ${message.userId} and highlight the most important parts to them. ` +
    'And if anyone is waiting on them for a response, highlight that prominently. ' +
    'Be concise, clear, direct, specific, and correct.';
  message.summary = await callLanguageModel(prompt);
}

/** Classifies and prioritizes a message via the language model. */
export async function classifyMessage(message: Message): Promise<void> {
  const prompt = `
        ## Message: ${JSON.stringify(message)}
        ----
        ## Prompt:
        Classify the email into one of the following categories with definitions:
        ${JSON.stringify(getClassificationDefinitions())}

        Then assign a priority level:
        ${JSON.stringify(getPriorityDefinitions())}

        Consider:
        - **Content:** Analyze the main topics and purpose.
        - **Tone:** Determine the formality and sentiment.
        - **Sender:** Identify if the sender is known or reputable.
        - **Time-Sensitivity:** Assess if immediate action is needed.
        - **Calls to Action:** Look for requests or required responses.


        RESPONSE FORMAT:
        Return JSON only: {"classification": "...", "priority": "...", "explanation": "..."}
        ----`;
  const response = await callLanguageModel(prompt);
  const parsed = markdownToJson(response) as Classification;

  // Add the Holoform prefix to the classification and priority.
  const classification: Classification = {
    classification: getFullLabel(parsed.classification),
    priority: getFullLabel(parsed.priority),
    explanation: parsed.explanation,
  };
  message.classification = classification;

  logger.debug(`\nMessage Classification: ${classification.classification}\n`);
  logger.debug(`Priority Level: ${classification.priority}\n`);
  logger.debug(`Explanation: ${classification.explanation}\n`);
}

/** Applies Holoform labels (and conditionally archives) based on classification. */
export function updateLabels(message: Message): void {
  const { classification, gmailMessageId, userId } = message;
  if (!classification) {
    throw new Error(`Cannot update labels: message ${gmailMessageId} is unclassified.`);
  }

  logger.debug(`Updating labels for message: ${gmailMessageId}`);
  const labelsToAdd = [
    getFullLabel(HoloformLabel.Holoform),
    classification.classification,
    classification.priority,
  ];
  let labelsToRemove: string[] = [];
  if (
    !(
      classification.classification.toLowerCase() === getFullLabel(HoloformLabel.ActionRequired) ||
      classification.priority === getFullLabel(HoloformLabel.HighPriority)
    )
  ) {
    labelsToRemove = [GmailSystemLabel.Inbox];
  }
  logger.debug(`\nLabels to add: ${labelsToAdd.join(', ')}\n`);
  logger.debug(`\nLabels to remove: ${labelsToRemove.join(', ')}\n`);
  changeEmailLabels({ emailId: gmailMessageId, labelsToAdd, labelsToRemove, userId });
  logger.debug(`Finished updating labels for message: ${gmailMessageId}\n`);
}

/** Drafts a reply-all response to the message via the language model. */
export async function draftResponse(message: Message): Promise<void> {
  const draftPrompt =
    `## Message: ${JSON.stringify(message)}\n` +
    `${'-'.repeat(8)}\n` +
    '## Prompt: \n' +
    'Draft a response to the message. ' +
    `Write in the first person as ${message.userId}. ` +
    'Be concise, clear, direct, specific, authentic, empathetic, personal, and correct. ' +
    'Use technical language where appropriate but remain plainspoken. Avoid clichés. ' +
    'The response should be a reply-all, sent to the sender and all other recipients. Retain CC. ' +
    "Reponse should be in JSON format: {'thread_id': '...', 'to': '...', 'cc': '...', 'subject': '...', 'body': '...'}";
  const draft = await callLanguageModel(draftPrompt);
  logger.debug(`draft_response_email: response draft: ${draft} \n`);
}

/** Processes a single message: fetch thread, classify, label. */
export async function processMessage(message: Message): Promise<Message> {
  message.threadMessages = getThreadMessageList(message.threadId, message.userId);
  await classifyMessage(message);
  logger.debug(
    `\n\n Subject: ${message.subject} \n Classification: ${JSON.stringify(message.classification)} \n ${'-'.repeat(20)}\n`,
  );
  updateLabels(message);
  logger.debug(`\n\nFinished processing message: ${message.gmailMessageId}\n\n`);
  message.status = 'processed';
  return message;
}

/** Processes every message. */
export async function processMessages(messages: Message[]): Promise<Message[]> {
  return Promise.all(messages.map(message => processMessage(message)));
}
